src/analysis/souffle/operations.dl

```diff
@@ -0,0 +1,80 @@
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_ANALYSIS_SOUFFLE_OPERATIONS_DL_
+#define SRC_ANALYSIS_SOUFFLE_OPERATIONS_DL_
+
+#include "dataflow_graph.dl"
+
+// The symbol representing the operator. For operations commonly represented by
+// a single character operator (such as +, =, /, *, etc), the operator will be
+// the tsring containing that character.
+.type Operator <: symbol
+
+// A list of operands, used for gathering all operators in a single relation
+// regardless of arity. Useful for rules that apply to any operation.
+.type OperandList = [ operand: AccessPath, next: OperandList ]
+
+// Operations of any arity go in this relation.
+.decl operation(owner: Principal, operator: Operator, result: AccessPath, operandList: OperandList)
+
+// unaryOperation and binaryOperation are convenience relations for the
+// very-common-cases of unary and binary operators.
+.decl unaryOperation(owner: Principal, operator: Operator, result: AccessPath, input1: AccessPath)
+.decl binaryOperation(
+  owner: Principal, operator: Operator, result: AccessPath, input1: AccessPath, input2: AccessPath)
+
+// Assign is syntactic sugar for unaryOperation(owner, "=", ...).
+.decl assign(owner: Principal, result: AccessPath, input1: AccessPath)
+
+// Cross-populate unaryOperation and operation.
+operation(owner, operator, result, [input1, nil]) :- unaryOperation(owner, operator, result, input1).
+unaryOperation(owner, operator, result, input1) :- operation(owner, operator, result, [input1, nil]).
+
+// Cross-populate binaryOperation and operation.
+operation(owner, operator, result, [input1, [input2, nil]]) :-
+  binaryOperation(owner, operator, result, input1, input2).
+binaryOperation(owner, operator, result, input1, input2) :-
+  operation(owner, operator, result, [input1, [input2, nil]]).
+
+// assign is a '=' operation with a single operand.
+unaryOperation(owner, "=", result, input1) :- assign(owner, result, input1).
+assign(owner, result, input1) :- unaryOperation(owner, "=", result, input1).
+
+// resolvedEdges are the same thing as assignments. Cross populate them to each
+// other.
+assign(owner, result, input1) :- resolvedEdge(owner, input1, result).
+resolvedEdge(owner, src, dst) :- assign(owner, dst, src).
+
+// Explode the operand lists so that we can extract the heads of each sublist
+// to get a mapping from full operations to their operands.
+.decl operationToPartialOperandLists(
+  owner: Principal, operator: Operator, result: AccessPath, operandList: OperandList,
+  partialList: OperandList)
+
+operationToPartialOperandLists(owner, operator, result, operandList, operandList) :-
+  operation(owner, operator, result, operandList).
+operationToPartialOperandLists(owner, operator, result, operandList, tail) :-
+  operationToPartialOperandLists(owner, operator, result, operandList, [_, tail]).
+
+// Map full operations to operands.
+.decl operationToOperands(
```

Comment:
Should this just be `operationToOperand(owner, operator, result, operand)`?

Reply:
My thinking was that I would lose information in the case that the same `result` was targeted by the result of two different `operation`s with the same `owner` and `operator` but different `operandList`s. That could cause ambiguity when showing a backtrace when we get around to implementing backtraces. That said, I'm using a wildcard everywhere I'm referring to this relation right now, so that might be just fine. Would be easy to add the `operandList` back in later if we needed it. I'll take `operandList` out for now.

Reply:
Done.